flink-ml-lib/src/main/java/org/apache/flink/ml/common/datastream/AllReduceUtils.java

```diff
@@ -0,0 +1,286 @@
+ * do reduce on a partition of the whole data and they all get the final reduce result. In detail,
+ * we split each double array into chunks of fixed size buffer (4KB by default) and let each subtask
+ * handle several chunks.
+ *
+ * <p>There're mainly three stages:
+ * <li>1. All workers send their partial data to other workers for reduce.
+ * <li>2. All workers do reduce on all data it received and then send partial results to others.
+ * <li>3. All workers merge partial results into final result.
+ */
+public class AllReduceUtils {
+
+    private static final int TRANSFER_BUFFER_SIZE = 1024 * 4;
+
+    /**
+     * Applies allReduce on the input data stream. The input data stream is supposed to contain one
+     * double array in each partition. The result data stream has the same parallelism as the input,
+     * where each partition contains one double array that sums all of the double arrays in the
+     * input data stream.
+     *
+     * <p>Note that we throw exception when one of the following two cases happen:
+     * <li>1. There exists one partition that contains more than one double array.
+     * <li>2. The length of double array is not consistent among all partitions.
+     *
+     * @param input The input data stream.
+     * @return The result data stream.
+     */
+    public static DataStream<double[]> allReduce(DataStream<double[]> input) {
+        // chunkId, totalElements, partitionedArray
+        DataStream<Tuple3<Integer, Integer, double[]>> allReduceSend =
+                input.flatMap(new AllReduceSend()).name("all-reduce-send");
+
+        // taskId, chunkId, totalElements, partitionedArray
+        DataStream<Tuple4<Integer, Integer, Integer, double[]>> allReduceSum =
+                allReduceSend
+                        .partitionCustom(
+                                (chunkId, numPartitions) -> chunkId % numPartitions, x -> x.f0)
+                        .transform(
+                                "all-reduce-sum",
+                                new TupleTypeInfo<>(
+                                        BasicTypeInfo.INT_TYPE_INFO,
+                                        BasicTypeInfo.INT_TYPE_INFO,
+                                        BasicTypeInfo.INT_TYPE_INFO,
+                                        PrimitiveArrayTypeInfo.DOUBLE_PRIMITIVE_ARRAY_TYPE_INFO),
+                                new AllReduceSum())
+                        .name("all-reduce-sum");
+
+        return allReduceSum
+                .partitionCustom((taskIdx, numPartitions) -> taskIdx % numPartitions, x -> x.f0)
+                .transform(
+                        "all-reduce-recv", TypeInformation.of(double[].class), new AllReduceRecv())
+                .name("all-reduce-recv");
+    }
+
+    /**
+     * Splits each double array into multiple chunks and send each chunk to the corresponding
+     * partition.
+     */
+    private static class AllReduceSend
```

Comment:
Every operation has `send` and `receive` semantics. Thus `AllReduceSend` does not convey much information of the key semantics here. Same for `AllReduceRecv`.

How about renaming this to e.g. `SplitArrayIntoChunks`? The other two operators might be renamed to e.g. `AggregateAndBroadcastChunks` and `JoinChunksIntoArray`.

Reply:
I think as we are implementing AllReduce, we should probably follow the existing AllReduce semantics[1], i.e., puting stuff in `sendBuf` and `recvBuf`.

[1] https://docs.microsoft.com/en-us/message-passing-interface/mpi-allreduce-function